File: libNative/src/main/java/com/munch/lib/libnative/exception/ExceptionCatchHandler.java
package com.munch.lib.libnative.exception;

import com.munch.lib.libnative.helper.AppHelper;

/**
 * Created by Munch on 2018/12/27 16:11.
 */
public abstract class ExceptionCatchHandler implements Thread.UncaughtExceptionHandler {

    private Thread.UncaughtExceptionHandler mExceptionHandler;

    public void init() {
        mExceptionHandler = Thread.getDefaultUncaughtExceptionHandler();
        Thread.setDefaultUncaughtExceptionHandler(this);
    }

    @Override
    public void uncaughtException(Thread t, Throwable e) {
        //未处理则让系统处理
        if (!handleException(t, e)) {
            mExceptionHandler.uncaughtException(t, e);
            return;
        }
        //自行处理完退出应用
        AppHelper.exit();
    }

    protected abstract boolean handleException(Thread t, Throwable e);
}

File: libfast/src/main/java/com/munch/lib/fast/base/ExceptionCatchHandler.kt
package com.munch.lib.fast.base

import android.os.Looper
import com.munch.lib.app.AppForegroundHelper
import com.munch.lib.app.AppHelper
import com.munch.lib.base.destroy
import com.munch.lib.fast.dialog.SimpleDialog
import com.munch.lib.helper.toDate
import com.munch.lib.log.Log2FileHelper
import com.munch.lib.log.log
import com.munch.lib.task.handler
import com.munch.lib.task.isMain
import java.io.File

/**
 * Create by munch1182 on 2021/11/24 15:49.
 */
object ExceptionCatchHandler : Thread.UncaughtExceptionHandler {

    private val dir by lazy { File(AppHelper.app.cacheDir, "/e") }
    private val log by lazy { Log2FileHelper(dir) }

    val exceptionFileDir: File
        get() = dir

    fun handle() {
        Thread.setDefaultUncaughtExceptionHandler(this)
    }

    override fun uncaughtException(t: Thread, e: Throwable) {
        log(e)
        if (t.isMain()) {
            /*thread(loop = true) { showDialog(t, e) }*/
            showDialog(t,e)
            Looper.loop()
        } else {
            handler { showDialog(t, e) }
        }
    }

    private fun showDialog(t: Thread, e: Throwable) {
        log.write("${System.currentTimeMillis().toDate()}:")
        log.write(e)
        AppForegroundHelper.currentActivity?.let {
            SimpleDialog(it)
                .setName("提示")
                .setContent("error: \n${e.localizedMessage}\n${t.name}")
                .setOnSureListener("重启") { d ->
                    d.cancel()
                    destroy()
                }
                .show(!t.isMain())
        }
    }
}